UI: Resize projector window by resolution or scale

fsworld009 opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Make sure you’ve read the contribution guidelines here: https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst -->

### Description
<!--- Describe your changes in detail. -->
<!--- If this change includes UI elements, please include screenshots. -->
Add `Resize Window` to Projector window's context menu to allow setting window size by resolution or scale.
Add a dialog when choosing `Custom...` to allow users to manually input desired resolution or scale.
![image](https://user-images.githubusercontent.com/7109839/203696564-af916612-e1da-46e4-aa8c-ea2722f56c0e.png)
![image](https://user-images.githubusercontent.com/7109839/203696570-85c2d5d3-6a93-40ce-b074-56e74094f960.png)
![image](https://user-images.githubusercontent.com/7109839/203696582-641a25cb-775e-4087-b036-40ea77795f52.png)


### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open GitHub Issue, or implements feature request -->
<!--- from the Ideas page, please link to the issue here. -->
There is an existing PR that trys to achieve resizing projector window by scale: #4378
However, that PR is put on hold due to some issues with integrating QDoubleSpinBox into context menu that require a custom QT component.

This PR changes UX by opening a standalone dialog window for custom input, to avoid the need of doing such task.

Additional changes compared to the above PR:
1. Add preset resolutions in context menu, which should be handy when previwing scenes or capture cards.
2. Present scales as percentage values, so that we can restrain users to only input integers, this should avoid the issue discussed in https://github.com/obsproject/obs-studio/pull/4378#discussion_r598067480
3. Since a standalone dialog is used for custom input, it can accept either resolution values or scale value.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment (hardware, OS version, etc.),-->
<!--- and the tests you ran, including how it may affect other areas of code. -->
Built the code on my Windows 10 Machine and manually tested my changes.

```
Edition	Windows 10 Pro
Version	21H2
Installed on	‎6/‎6/‎2020
OS build	19044.2251
Experience	Windows Feature Experience Pack 120.2212.4180.0
```

### Types of changes
<!--- What types of changes does your PR introduce? Uncomment all that apply -->
<!--- - Bug fix (non-breaking change which fixes an issue) -->
- New feature (non-breaking change which adds functionality)
<!--- - Tweak (non-breaking change to improve existing functionality) -->
<!--- - Performance enhancement (non-breaking change which improves efficiency) -->
<!--- - Code cleanup (non-breaking change which makes code smaller or more readable) -->
<!--- - Breaking change (fix or feature that would cause existing functionality to change) -->
<!--- - Documentation (a change to documentation pages) -->

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [ ] I have included updates to all appropriate documentation.

fsworld009 commented:
> I have included updates to all appropriate documentation.

There seems to be no existing documentation about the projector window, so I'm not sure if there's any TODO required for this one. (I also don't see any doc change in #4378)

Warchamp7 reviewed (changes requested):
Hi @fsworld009, apologies for my delay reviewing this. I think this is a much more solid approach for this feature, thanks for the hard work! I have a couple requests and questions for you.

**- Build this window as an actual UI form rather than entirely in code**
This is preferred for anything more complex than a dialog or that is constructed programmatically


**- Use QDialogButtonBox for the Ok dialog button at the bottom**
Add in the usual Cancel button for discarding the action as well


**- Add labels for the input fields**
Ex. `Resize Method` for the Resolution/Scale dropdown. `Size` for the custom input


**- For custom scale, include the % symbol inline in the input**
You can see an example of this with the bitrate and keyframe interval fields in encoder settings
![image](https://user-images.githubusercontent.com/1554753/220145562-6768adb9-093d-4103-b855-7dd54bfd8649.png)

**- For custom resolution, use an editable QComboBox like we do in Video settings**
![image](https://user-images.githubusercontent.com/1554753/220147886-74a728fe-e4aa-4f24-b29b-ba5ebdcdde94.png)

These changes will help this dialog look and behave similar to others in the program.

-----

As an extra question and request, is it possible to detect the resolution of the screen the projector is on? If so can you hide preset options from the menu if they are larger than the screen?

Ex. If the projector is on a 1920x1080 screen, hide 2560x1440 and 3840x2160. Users can still use Custom if they want those sizes.

The exception here should be for 100% scale, which should always be available, even when the source being projectored is larger than the screen.

fsworld009 commented:
## 08.14.2023 Update

(New changes are only tested in Windows 10 for now)

Rebased to resolve conflicts and updated:
1. Redid dialog UI with `.ui` Qt design file and followed existing settings UI layout.
   1. Input label is added
   2. Resolution input is now using the same format as in video settings
   3. Scale input now has suffix in the input
2. Updated Context menu preset items: only resolutions or scales lower than screen resolution will be populated (except 100%, which is always available)

New dialog screenshots
![0814-1](https://github.com/obsproject/obs-studio/assets/7109839/364f6b49-d9ca-46b0-9a4d-51b21c1d10b2)
![0814-2](https://github.com/obsproject/obs-studio/assets/7109839/85f4f759-bec1-46bd-aa03-1db7b8a2b47c)
![0814-5](https://github.com/obsproject/obs-studio/assets/7109839/cfc62857-865e-43e6-8709-e64384ab522e)

Context menu presets
For a 1920x1080 source, when displayed in 4K screen:
![0814-3](https://github.com/obsproject/obs-studio/assets/7109839/1f3f7ab5-acd9-42ac-bcd3-d1af8091ede0)

when displayed in 1080p screen:
![0814-4](https://github.com/obsproject/obs-studio/assets/7109839/5f84b737-662a-4636-83ad-4711405a0a6e)

fsworld009 commented:
@Warchamp7 please help rereview when you get a chance, thank you.

PatTheMav commented:
@fsworld009 Does your code take DPI scaling into account?

fsworld009 commented:
@Scrxtchy good question, I personally don't use scaling so didn't try it. What would be the expected behavior?

fsworld009 commented:
@PatTheMav good question, I personally don't use scaling so didn't try it. What would be the expected behavior?

fsworld009 commented:
> @fsworld009 Does your code take DPI scaling into account?

@PatTheMav according to Qt6 [documentation ](https://doc.qt.io/qt-6/highdpi.html#conceptual-model) the DPI scaling from OS should be automatically applied. I tried on Windows and confirmed the behavior as well.

fsworld009 commented:
Updated:

1. Squashed and rebased from latest master in order to resolve conflicts caused by file renames
2. Renamed files and updated header include paths based on current file structure in master branch
3. Some changes caused by running clang format after rebase (I assume this is because the rule has been updated at least once after the PR was opened).

The current commit is based on [08.14.2023 update](https://github.com/obsproject/obs-studio/pull/7851#issuecomment-1677416480) which is based on @Warchamp7 's [guideline](https://github.com/obsproject/obs-studio/pull/7851#pullrequestreview-1306012561) for this PR.

@Warchamp7 please review this PR when you have a chance, thanks.

fsworld009 commented on frontend/widgets/OBSProjector.cpp:
```diff
@@ -453,6 +459,137 @@ void OBSProjector::ResizeToContent()
+
+std::vector<std::pair<int, int>> OBSProjector::GetResizeResolutionPresets()
+{
+	int resolutionPresets[][2] = {{1280, 720}, {1920, 1080}, {2560, 1440}, {3840, 2160}};
```
Sorry I'm not sure what is the expectation you are looking for. Are you suggesting we should remove all fixed resolution presets, and expanding percentage preset list up to 400%? For example, a list of 100%, 125%, 150%, 200%, 250%, 300%, 350%, 400%?
